libp2p/switch.nim

```diff
@@ -417,30 +426,48 @@ proc mount*[T: LPProtocol](s: Switch, proto: T, matcher: Matcher = nil) {.gcsafe
   s.ms.addHandler(proto.codecs, proto, matcher)
 
 proc accept(s: Switch, transport: Transport) {.async.} = # noraises
-  ## transport's accept loop
+  ## switch accept loop, ran for every transport
   ##
 
+  let upgrades = AsyncSemaphore.init(ConcurrentUpgrades)
   while transport.running:
     var conn: Connection
     try:
       debug "About to accept incoming connection"
+      await upgrades.acquire()
       conn = await transport.accept()
-      if not isNil(conn):
-        debug "Accepted an incoming connection", conn
-        asyncSpawn s.upgradeIncoming(conn) # perform upgrade on incoming connection
-      else:
+      if isNil(conn):
         # A nil connection means that we might have hit a
         # file-handle limit (or another non-fatal error),
         # we can get one on the next try, but we should
         # be careful to not end up in a thigh loop that
         # will starve the main event loop, thus we sleep
         # here before retrying.
+        trace "Unable to get a connection, sleeping"
         await sleepAsync(100.millis) # TODO: should be configurable?
+        upgrades.release()
+        continue
+
+      debug "Accepted an incoming connection", conn
+      proc upgradeMonitor() {.async.} =
+        ## monitor connection for upgrades
+        ##
+        try:
+          await conn.upgraded.wait(30.seconds) # wait for connection to by upgraded
```

Comment:
so this implicitly becomes an upgrade timeout? like this overlaps with the general connection timeout monitor, creating two competing timeout mechanisms

Reply:
There is no upgrade timeout right now, so yes this is an implicit upgrade timeout.

Reply:
so it looks like a feature that should at least be shared between incoming and outgoing upgrades?

Reply:
Well, this is actually more about who initiates the flow. In the case of a `dial/connect`, "we" are the initiator and we can make decisions when to stop it - either due to a timeout or some other reasons. In other words "we" control the flow of the upgrade. For incoming connections, we're not the initiator, the remote is, we only respond to it's upgrade requests, so this is a guard to prevent it from hijacking a connection in case it "hangs".

Reply:
fair enough, deserves a comment